subsys/bluetooth/audio/has_client.c

```diff
@@ -82,11 +88,179 @@ static enum bt_has_capabilities get_capabilities(const struct has_inst *inst)
+	}
+
+	record.index = pdu->index;
+	record.properties = pdu->properties;
+	record.name = name;
+
+	name_len = buf->len + 1; /* + 1 byte for NULL terminator */
+	if (name_len > ARRAY_SIZE(name)) {
+		BT_WARN("name is too long (%zu > %u)", buf->len, BT_HAS_PRESET_NAME_MAX);
+
+		name_len = ARRAY_SIZE(name);
+	}
+
+	utf8_lcpy(name, pdu->name, name_len);
+
+	client_cb->preset_read_rsp(&inst->has, 0, &record, !!pdu->is_last);
+}
+
+static void handle_generic_update(struct has_inst *inst, struct net_buf_simple *buf, bool is_last)
+{
+	const struct bt_has_cp_generic_update *pdu;
+	struct bt_has_preset_record record;
+	char name[BT_HAS_PRESET_NAME_MAX + 1]; /* + 1 byte for null-terminator */
+	size_t name_len;
+
+	if (buf->len < sizeof(*pdu)) {
+		BT_ERR("malformed PDU");
+		return;
+	}
+
+	pdu = net_buf_simple_pull_mem(buf, sizeof(*pdu));
+
+	record.index = pdu->index;
+	record.properties = pdu->properties;
+	record.name = name;
+
+	name_len = buf->len + 1; /* + 1 byte for NULL terminator */
+	if (name_len > ARRAY_SIZE(name)) {
+		BT_WARN("name is too long (%zu > %u)", buf->len, BT_HAS_PRESET_NAME_MAX);
+
+		name_len = ARRAY_SIZE(name);
+	}
+
+	utf8_lcpy(name, pdu->name, name_len);
+
+	client_cb->preset_update(&inst->has, pdu->prev_index, &record, is_last);
+}
+
+static void handle_preset_deleted(struct has_inst *inst, struct net_buf_simple *buf, bool is_last)
+{
+	if (buf->len < sizeof(uint8_t)) {
+		BT_ERR("malformed PDU");
+		return;
+	}
+
+	client_cb->preset_deleted(&inst->has, net_buf_simple_pull_u8(buf), is_last);
+}
+
+static void handle_preset_availability(struct has_inst *inst, struct net_buf_simple *buf,
+				       bool available, bool is_last)
+{
+	if (buf->len < sizeof(uint8_t)) {
+		BT_ERR("malformed PDU");
+		return;
+	}
+
+	client_cb->preset_availability(&inst->has, net_buf_simple_pull_u8(buf), available,
+				       is_last);
+}
+
+static void handle_preset_changed(struct has_inst *inst, struct net_buf_simple *buf)
+{
+	const struct bt_has_cp_preset_changed *pdu;
+
+	BT_DBG("conn %p buf %p", (void *)inst->conn, buf);
+
+	if (buf->len < sizeof(*pdu)) {
+		BT_ERR("malformed PDU");
+		return;
+	}
+
+	pdu = net_buf_simple_pull_mem(buf, sizeof(*pdu));
+
+	if (pdu->is_last > BT_HAS_IS_LAST) {
+		BT_WARN("unexpected is_last 0x%02x", pdu->is_last);
+	}
+
+	switch (pdu->change_id) {
+	case BT_HAS_CHANGE_ID_GENERIC_UPDATE:
+		if (client_cb->preset_update) {
+			handle_generic_update(inst, buf, !!pdu->is_last);
+		}
+		break;
+	case BT_HAS_CHANGE_ID_PRESET_DELETED:
+		if (client_cb->preset_deleted) {
+			handle_preset_deleted(inst, buf, !!pdu->is_last);
+		}
+		break;
+	case BT_HAS_CHANGE_ID_PRESET_AVAILABLE:
+		if (client_cb->preset_availability) {
+			handle_preset_availability(inst, buf, !!pdu->is_last, true);
+		}
+		return;
+	case BT_HAS_CHANGE_ID_PRESET_UNAVAILABLE:
+		if (client_cb->preset_availability) {
+			handle_preset_availability(inst, buf, !!pdu->is_last, false);
+		}
+		return;
```

Comment:
@Thalley Ah, sorry. Here ^^ So the notification is not handled if the callback is not set

Reply:
Would it make more sense to move the `if (client_cb->preset_availability) {` to the `handle_preset_availability` function? Keeping this `switch` more "clean" and also avoiding having `if (client_cb->preset_availability) {` twice her?